:construction:Extra optimizations for a future firmware

leptun opened the pull request:
This PR is composed of some optimizations which were left out of #3365. This PR is simply so that we don't lose track of these commits. There is no plan for now to merge this PR in 3.12.0

@gudnimg fyi

gudnimg committed:
Optimise calculations to use hypot() where possible

It is defined: hypot(x,y) = sqrtf(x*x + y*y)

Changes save 122 bytes of flash memory

gudnimg committed:
Optimise calculations to use hypot() where possible

It is defined: hypot(x,y) = sqrtf(x*x + y*y)

Changes save 122 bytes of flash memory

leptun committed:
lcd avoid streams

leptun committed:
lcd avoid streams

leptun committed:
Avoid streams in general

No efficiency gained

leptun committed:
Avoid streams in general

No efficiency gained

leptun committed:
Tighter LCD timing

Saves 14B of flash

leptun committed:
Tighter LCD timing

Saves 14B of flash

gudnimg commented:
Regarding the hypot commit. I will need to look at it from assembly level. It will take time. From first glance hypot seems to be slower, but I need to break down the subroutines to be sure.

Another note, I think we can convert the MarlinSerial ring buffer elements to 1 byte. (We are converting them to 2 bytes in places)

leptun commented:
We can profile those functions using the _micros as a timebase. I don't expect it to be _that_ much slower, but any speed penalty is not really welcome in the mbl code. It's already struggling right now when you push the stepper isr to the limit.
I also want to make sure the stepping is identical between implementations. If that is the case, then I see no issue with hypot.

I agree, we can change the ring_buffer to use uint8_t. We will also merge ring_buffer with rbuf at some point. One is used for UART0/1 and the other for UART2(MMU). We can use a common implementation, but for the purpose of this PR we should keep them split.
We could also optimize this further by making sure RX_BUFFER_SIZE is a multiple of 2 and that it's in the 1-256 range. This way we can use & instead of % for the index arithmetic

Commit:
Optimise if statement in pid_extruder()

Saves 28 bytes of flash memory (Arduino IDE 1.8.19)

Commit:
Optimise if statement in pid_extruder()

Saves 28 bytes of flash memory (Arduino IDE 1.8.19)

leptun committed:
better lcd printing for IP address

leptun committed:
better lcd printing for IP address

gudnimg committed:
Remove unused variable

gudnimg committed:
Remove unused variable

gudnimg committed:
Remove unused variable z_endstop_invert

This also renders the function invert_z_endstop useless so I removed it as well.

I see no change in memory footprint but this may be considered code size optimisation.

gudnimg committed:
Remove unused variable z_endstop_invert

This also renders the function invert_z_endstop useless so I removed it as well.

I see no change in memory footprint but this may be considered code size optimisation.

Commit:
tmp_extruder should not expect a float value

tmp_extruder is uint8_t and not float

Saves 6 bytes of flash memory

Commit:
tmp_extruder should not expect a float value

tmp_extruder is uint8_t and not float

Saves 6 bytes of flash memory

gudnimg commented:
Found this code is not used. `MMU_M600_SWITCH_EXTRUDER` is not defined anywhere and the function `choose_extruder_menu()` doesn't exit. Do you recognise this code @leptun ?
![image](https://user-images.githubusercontent.com/8218499/153747456-25a2d064-71e6-4058-8c21-38bc5f902332.png)

leptun commented:
It's dead SNMM (MMU1) code. It will be removed soon

Commit:
heating_status_counter is only used in ultralcd.cpp so lets make it static there

Commit:
heating_status_counter is only used in ultralcd.cpp so lets make it static there

Commit:
Remove unnused variable saved_filament_type

saved_filament_type is never set, but because it is global it is initalised to 0 by default. The default filament type in prusa_statistics is 0 so we can remove this variable without impacting anything.

Changes save 54 bytes of flash memory

Commit:
Remove unnused variable saved_filament_type

saved_filament_type is never set, but because it is global it is initalised to 0 by default. The default filament type in prusa_statistics is 0 so we can remove this variable without impacting anything.

Changes save 54 bytes of flash memory

Commit:
cur_mA in M905 is expected to be float

tmc2130_cur2val input parameter is float, and I don't see why the we should be restricted to long type.

This actually saves 4 bytes of flash, was not expecting that.

Commit:
cur_mA in M905 is expected to be float

tmc2130_cur2val input parameter is float, and I don't see why the we should be restricted to long type.

This actually saves 4 bytes of flash, was not expecting that.

Commit:
Optimise TMC2130_SERVICE_CODES_M910_M918 code

Also fixed indentation and magic numbers in the lines I optimised.

Save 56 bytes of flash if TMC2130_SERVICE_CODES_M910_M918 is defined

Commit:
Optimise TMC2130_SERVICE_CODES_M910_M918 code

Also fixed indentation and magic numbers in the lines I optimised.

Save 56 bytes of flash if TMC2130_SERVICE_CODES_M910_M918 is defined

Commit:
Optimise M106 in restore_print_from_eeprom

fan_speed_rec ranges from 0 to 255. itostr3(fan_speed_rec) = fan_speed_rec thus this call to itostr3 is not needed.

Changes save 18 bytes of flash memory

Commit:
Optimise M106 in restore_print_from_eeprom

fan_speed_rec ranges from 0 to 255. itostr3(fan_speed_rec) = fan_speed_rec thus this call to itostr3 is not needed.

Changes save 18 bytes of flash memory

gudnimg commented:
Took a look at `hypot()` commit. Measured the clock ticks with Timer 1 at 16MHz.

I start by first measuring the microseconds of the call `delayMicroseconds(1000)` just to see if my calculations are good. I then measure how many ticks it takes to calculate `sqrt(x * x + y * y)` where `x = 1000.2` and `y = 50.5`. Then calculate `hypot(x, y)` and check if the results are the same to 10 decimal places. Then I finally measure `delayMicroseconds(1000)` again to verify there hasn't been any drift.

Note that I assign the result from `sqrt(x * x + y * y)` and `hypot(x, y)` to a variable, that will have an effect on the tick measurement.

I see no loss in precision in my example and don't see much different in clock cycles. So I would say these two are basically the same.

Below are the results:

![image](https://user-images.githubusercontent.com/8218499/153757266-c16d343e-d8bc-4da4-b4d6-c464509a1a78.png)

Arduino sketch with source code if anyone is interested:


<details>
<summary> Source Code </summary>
<br>

```cpp
float x;
float y;
volatile byte overflows = 0;

void setup()  
{ 
  Serial.begin(115200);
  while(!Serial);
  x = 1000.2;
  y = 50.5;
  float res = 0; // variable to force the program to store math result to memory

  Serial.println("|----------------- Verify 1000us Measurement -------------------|");
  Serial.flush(); // Make sure to flush so tick measurement is not affected by slow Serial stream
  
  TCCR0A = 0; // Disable Timer 0
  TCCR0B = 0; // Disable Timer 0
  TCCR1A = 0; // Set Timer 1 Control Register A to 0 
  TCCR1B = 0; // Set Timer 1 Control Register B to 0 
  TCCR1C = 0; // Set Timer 1 Control Register C to 0
  TIMSK1 = _BV(TOIE1); // Enable Time 1 overflow interrupt  

  TCNT1 = 0x0000; // Make sure tick count is 0
  TCCR1B = 1;  // Start Timer 1 with no prescaling (CLK = 16MHz)
  delayMicroseconds(1000); // Code being measured
  TCCR1B = 0;  // Stop Timer 1

  unsigned int temp_timer = TCNT1;  // Store ticks
  // Correct ticks in case there was overflow. -1 Here because ISR is called at start-up
  unsigned long ticks = ((((unsigned long)overflows - 1) << 16) | (unsigned long)temp_timer) - 4;

  Serial.print("Timer1 Ticks: ");
  Serial.println(temp_timer);
  Serial.print("Timer1 Overflows: ");
  Serial.println(overflows - 1);
  Serial.print("Timer1 Ticks after correcting for Overflows: ");
  Serial.println(ticks);

  Serial.print("Timer1 Ticks converted to microseconds: ");
  Serial.print((float)ticks/(F_CPU / 1000000), 4);
  Serial.println("us");

  Serial.println("|------------------ Measurement 1 -------------------|");
  Serial.flush();

  overflows = 0;       // Reset overflow counter
  TCNT1 = 0x0000;      // Make sure tick count is 0
  TCCR1B = 1;          // Start Timer 1 with no prescaling (CLK = 16MHz)
  res = sqrt(x*x+y*y); // Code being measured
  TCCR1B = 0;          // Stop Timer 1

  temp_timer = TCNT1;  // Store ticks
  // Correct ticks in case there was overflow
  ticks = ((((unsigned long)overflows) << 16) | (unsigned long)temp_timer) - 4;

  Serial.print("sqrt(x*x+y*y) = ");
  Serial.println(res, 10);
  Serial.print("Timer1 Ticks: ");
  Serial.println(temp_timer);
  Serial.print("Timer1 Overflows: ");
  Serial.println(overflows);
  Serial.print("Timer1 Ticks after correcting for Overflows: ");
  Serial.println(ticks);
  Serial.print("Timer1 Ticks converted to microseconds: ");
  Serial.print((float)ticks/(F_CPU / 1000000), 4);
  Serial.println("us");

  Serial.println("|------------------ Measurement 2 -------------------|");
  Serial.flush();

  overflows = 0;     // Reset overflow counter
  TCNT1 = 0x0000;    // Make sure tick count is 0
  TCCR1B = 1;        // Start Timer 1 with no prescaling (CLK = 16MHz)
  res = hypot(x, y); // Code being measured
  TCCR1B = 0;        // Stop Timer 1

  temp_timer = TCNT1; // Store ticks
  // Correct ticks in case there was overflow
  ticks = ((((unsigned long)overflows) << 16) | (unsigned long)temp_timer) - 4;

  Serial.print("hypot(x, y) = ");
  Serial.println(res, 10);
  Serial.print("Timer1 Ticks: ");
  Serial.println(temp_timer);
  Serial.print("Timer1 Overflows: ");
  Serial.println(overflows);
  Serial.print("Timer1 Ticks after correcting for Overflows: ");
  Serial.println(ticks);
  Serial.print("Timer1 Ticks converted to microseconds: ");
  Serial.print((float)ticks/(F_CPU / 1000000), 4);
  Serial.println("us");

  Serial.println("|----------------- Verify 1000us Measurement again -------------------| ");
  Serial.flush();

  overflows = 0;           // Reset overflow counter
  TCNT1 = 0x0000;          // Make sure tick count is 0
  TCCR1B = 1;              // Start Timer 1 with no prescaling (CLK = 16MHz)
  delayMicroseconds(1000); // Code being measured
  TCCR1B = 0;              // Stop the timer

  temp_timer = TCNT1;      // Store ticks
  // Correct ticks in case there was overflow
  ticks = ((((unsigned long)overflows) << 16) | (unsigned long)temp_timer) - 4;

  Serial.print("Timer1 Ticks: ");
  Serial.println(temp_timer);
  Serial.print("Timer1 Overflows: ");
  Serial.println(overflows);
  Serial.print("Timer1 Ticks after correcting for Overflows: ");
  Serial.println(ticks);
  Serial.print("Timer1 Ticks converted to microseconds: ");
  Serial.print((float)ticks/(F_CPU / 1000000), 4);
  Serial.println("us");
  Serial.flush();
  abort();
}

void loop() {}

ISR(TIMER1_OVF_vect)   
{  
  overflows++;  
}//end ISR 
```

</details>

leptun commented:
If you need to do measurement that requires clock precision in Prusa-Firmware (eg: measuring how long something takes in an ISR), you can do this very easily with TIM5. Beware that you need to use TIM3 on the miniRambo and that it has NO overflow handling.

At the end of setup:
```
    TIMSK5 = 0;
    TCCR5A = 0;
    TCCR5B = 1;
```
Where you want to measure:
```
    uint16_t initial = TCNT5;
    stepper_check_endstops(); //whatever function you want to measure
    printf_P(PSTR("time:%u\n"), TCNT5 - initial);
```
If a longer period is needed, increase TCCR5B (it increases the prescaler) and adjust your math accordingly. See datasheet for the prescaler values.


It is good that the code behaves correctly. I just hope it doesn't have any unforseen side effects.

gudnimg committed:
Remove unused variable maxlimit_status

gudnimg committed:
Remove unused variable maxlimit_status

gudnimg committed:
ftostr12ns: change xx from long to int

Largest expected number is 999 after the multiplication by 100.

Changes save 44 byte of flash memory. I measured the execution time drops from ~170us to ~73us.

gudnimg committed:
ftostr12ns: change xx from long to int

Largest expected number is 999 after the multiplication by 100.

Changes save 44 byte of flash memory. I measured the execution time drops from ~170us to ~73us.

gudnimg committed:
Use set_destination_to_current instead of memcpy

Saves 14 bytes of flash memory

gudnimg committed:
Use set_destination_to_current instead of memcpy

Saves 14 bytes of flash memory

gudnimg committed:
Reduce a few magic numbers for FILENAME_LENGTH

gudnimg committed:
Reduce a few magic numbers for FILENAME_LENGTH

gudnimg committed:
Remove one redundant st_synchronize() call

st_synchronize() is called just before leaving gcode_G28() so we don't need to call it again just after leaving gcode_G28()

This saves 4 bytes of flash and some clock cycles :)

gudnimg committed:
Remove one redundant st_synchronize() call

st_synchronize() is called just before leaving gcode_G28() so we don't need to call it again just after leaving gcode_G28()

This saves 4 bytes of flash and some clock cycles :)

gudnimg commented:
I have a feeling `buflen` in `cmdqueue.cpp` can be `uint8_t` instead of `int`. Even if all the commands in the buffer are 3 letters, Mxx or Gxx, that would still only fit 164 commands/lines. The command buffer only holds 493 bytes 🤔 

Highest I've seen so far for `buflen` is 16. I may look more into this next weekend as this can shave off 120 bytes of flash if possible.

leptun commented:
Please let's not touch the cmdqueue buffers for now. It works through misterious ways and I don't want to break anything there. A lot of the older lcd code uses it and I'm afraid we might break it quite easily accidentally.

gudnimg commented:
@leptun I noticed `endstops_hit_on_purpose()` function is not doing anything useful for us. We don't read it's return value anywhere we use it so it is just wasting clock cycles imo. Do you agree we should remove it? It saves 58 bytes of flash on my end.
https://github.com/prusa3d/Prusa-Firmware/blob/b654217a5b3d77816eaeeb5ea803cc41de5e9b74/Firmware/stepper.cpp#L228

Edit: Or instead of removing it, we remove the return value and just keep `endstop_hit = 0`. But then `checkHitEndstops()` can take its place I guess.

leptun commented:
I guess we can make endstops_hit_on_purpose() void and just clear endstop_hit. Please don't change it more than that

leptun commented:
@gudnimg Off topic question, do you have a MK3 or a MK2.5 with the PAT9125 filament sensor?

gudnimg commented:
I only have the MK3S+ printer

gudnimg committed:
Combine two calls to lcd_clear()

In this if-else statement lcd_clear is always called. We can pull the call outside the if-else clause.

Saves 4 bytes of flash

gudnimg committed:
Combine two calls to lcd_clear()

In this if-else statement lcd_clear is always called. We can pull the call outside the if-else clause.

Saves 4 bytes of flash

gudnimg committed:
Remove more redundant calls to lcd_clear()

lcd_clear() is not needed before lcd_return_to_status()
Because  lcd_return_to_status() will later call lcd_clear()

Changes save 12 bytes of flash memory

gudnimg committed:
Remove more redundant calls to lcd_clear()

lcd_clear() is not needed before lcd_return_to_status()
Because  lcd_return_to_status() will later call lcd_clear()

Changes save 12 bytes of flash memory

leptun commented:
@gudnimg make sure you don’t change something in the recently deleted snmm code. We should probably rebase this branch to the latest Mk3 to avoid this problem.

gudnimg commented:
@leptun Indeed, I'm trying to use the current MK3 as reference before switching to this branch. I'm really happy you removed the SNMM code 😄 

To add to my previous comment from 2 days ago... I got a MMU2S yesterday. So I will hopefully soon be able to test the MMU code directly 💪

leptun commented:
That’s great! Btw, see: https://github.com/prusa3d/Prusa-Firmware/issues/1876#issuecomment-1045965924

gudnimg commented:
@leptun I'm hesitant to run the rebase. Do you remember which git command you used before? I've usually used this guide https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a to update my fork.

leptun commented:
@gudnimg I use gitkraken for everything git (unless that is not working, in which case I just use git directly). This is how I did the rebase above. I highly recommend you try this as well. There is a free version available that can be used for almost everything except private repos.

https://user-images.githubusercontent.com/17808203/154795545-9045ce25-6aef-45e9-a58a-2e79fe51abb0.mp4

As for why it's so zoomed-in, it's because I use it on a larger display normally and it looks correctly there. I moved it to my secondary screen for the screen recording so that the file size would be smaller, so I wouldn't have to compress the video

gudnimg committed:
Optimise lcd_menu_show_sensors_state()

menu_lcd_lcdupdate_func() takes care of starting lcd_timeoutToStatus if the knob is clicked.

When the knob is clicked we only want to back out of the menu.
This changes makes the menus slightly more conistant behind the scenes.

Changes saves 22 bytes of flash memory.

gudnimg committed:
Optimise lcd_menu_show_sensors_state()

menu_lcd_lcdupdate_func() takes care of starting lcd_timeoutToStatus if the knob is clicked.

When the knob is clicked we only want to back out of the menu.
This changes makes the menus slightly more conistant behind the scenes.

Changes saves 22 bytes of flash memory.

gudnimg committed:
lcd_babystep_z should not start lcd_timeoutToStatus

When the menu is entered and left the function
menu_lcd_lcdupdate_func() takes care of resetting the timer.

Currently the firmware will reset the timer twice when the lcd_babystep_z menu is entered. This commit fixes that.

Changes save 8 bytes of flash memory

gudnimg committed:
lcd_babystep_z should not start lcd_timeoutToStatus

When the menu is entered and left the function
menu_lcd_lcdupdate_func() takes care of resetting the timer.

Currently the firmware will reset the timer twice when the lcd_babystep_z menu is entered. This commit fixes that.

Changes save 8 bytes of flash memory

gudnimg committed:
Optimise if statment

Saves 2 bytes of flash

gudnimg committed:
Optimise if statment

Saves 2 bytes of flash

gudnimg committed:
Optimise if statements in lcd_tune_menu

I am basically combining two if statements if(SilentModeMenu == SILENT_MODE_NORMAL)

Saves 10 bytes of flash

gudnimg committed:
Optimise if statements in lcd_tune_menu

I am basically combining two if statements if(SilentModeMenu == SILENT_MODE_NORMAL)

Saves 10 bytes of flash

gudnimg commented:
Build fails. It looks like to me it is a problem in the current CMake script https://github.com/prusa3d/Prusa-Firmware/blob/MK3/CMakeLists.txt It doesn't know where to find `"Configuration_prusa.h"`

<img width="667" alt="image" src="https://user-images.githubusercontent.com/8218499/155035636-ab1d1ab9-3f85-4ff4-abe5-7cd6f43e4aac.png">

leptun commented:
Yes. You can't include "Configuration_prusa.h" in anything that has "unit tests" (loosely used word here). Please discard any changes to mmu related code (such is the case with the auto-deplete). This will be addressed later.

gudnimg committed:
Remove unused variable bFilamentFirstRun

Saves 1 byte of SRAM

gudnimg committed:
Remove unused variable bFilamentFirstRun

Saves 1 byte of SRAM

gudnimg committed:
Remove one call to dtostrf

I could not see any difference in the functionality when testing this locally

Changes save 818 bytes of flash memory
(I suspect this is due to dtostrf no longer being called in the firmware for MK3S?)

gudnimg commented:
@leptun Managed to somehow create 818 byte reduction in flash. This seems almost too good to be true. Do you see any issues I might be missing? https://github.com/prusa3d/Prusa-Firmware/pull/3378/commits/0b198bddf6c678f6adff8e043f08f52d8ebdf525

I tested this locally by changing the format of `myfloat`. I did not see any difference no matter how many digits/decimals there are or if the value is negative.

```c
  float myfloat = 123.6789;
  char buf[25] = {0};
  Serial.print("dtostrf() ->");
  sprintf_P(buf, PSTR("G92 E"));
  dtostrf(myfloat, 6, 3, buf + strlen(buf));
  Serial.println(buf);
  char buf2[25] = {0};
  Serial.print("sprintf() ->");
  sprintf_P(buf2,PSTR("G92 E%6.3f"),myfloat);
  Serial.println(buf2);
```

leptun commented:
@gudnimg I'd like to make a "PR3" of optimizations where we cherrypick changes from this one. I'd like to go for the smallest changes that have the largest flash impact (such as dtostrf).

gudnimg committed:
Optimise strcpy calls to use strcpy_P

Saves 24 bytes of SRAM

leptun commented on Firmware/ultralcd.cpp:
```diff
@@ -1031,11 +1031,11 @@ void lcd_commands()
+			strcat_P(cmd1, PSTR(" I"));
 			strcat(cmd1, ftostr32(_Ki));
-			strcat(cmd1, " D");
+			strcat_P(cmd1, PSTR(" D"));
```
Maybe this can be turned into a single sprintf_P. It's just building a command using the resulting PID values. Same for M303 above.

gudnimg replied:
I thought about it but didn't have time to test it. `ftostr32` limits the string width, but `sprintf` wouldn't truncate the length.  

Edit: On second thought perhaps `snprintf_P` could work in that regard.

<img width="392" alt="image" src="https://user-images.githubusercontent.com/8218499/155842180-cee7fb20-fb4f-4fd4-9683-c1ed29f50025.png">

> Like [sprintf()](https://www.nongnu.org/avr-libc/user-manual/group__avr__stdio.html#ga6017094d9fd800fa02600d35399f2a2a), but instead of assuming s to be of infinite size, no more than n characters (including the trailing NUL character) will be converted to s.

> Returns the number of characters that would have been written to s if there were enough space.

gudnimg replied:
Another note, slightly related to `ftostr32`. I've found that the code in general in `conv2str.cpp` can be optimised but 80% of the code is unused so perhaps we can just replace it by being a bit more smart with AVR's stdio library https://www.nongnu.org/avr-libc/user-manual/group__avr__stdio.html

gudnimg committed:
Use sprintf_P for M301 instead of ftostr32

Also fixed indentation of if statement

Saves 248 bytes of flash

leptun committed:
Temperature selection is int16_t. Use sprintf for PID start command.

Saves 90B of flash and 2B of ram

gudnimg committed:
Purge ftostr3

Saves 54 bytes of Flash and 2 bytes of SRAM

gudnimg commented:
I replaced all `ftostr3` calls with `lcd_printf_P`. I didn't see memory savings until all `ftostr3` were replaced.

leptun commented:
Well, that makes sense. sprintf was also a function call, but by replacing all of them the ftostr3 was no longer included during linking.

gudnimg committed:
Optimise a few uses of SERIAL_PROTOCOLLN

Saves 34 bytes of flash memory and reduces code by 8 lines.

leptun commented on Firmware/Marlin_main.cpp:
```diff
@@ -8321,9 +8321,9 @@ SERIAL_PROTOCOLPGM("\n\n");
 			int16_t usteps = 0;
-			cal_status ? SERIAL_PROTOCOLLN("PINDA cal status: 1") : SERIAL_PROTOCOLLN("PINDA cal status: 0");
+			SERIAL_PROTOCOL("PINDA cal status: ");
+			SERIAL_PROTOCOLLN(calibration_status_pinda());
```
ummm... can't this be the PGM variant?
So `SERIAL_PROTOCOLPGM("PINDA cal status: ");`

gudnimg replied:
I think so. We can save a lot of SRAM (~50 bytes) but the flash usage will increase about 60 bytes if I remember correctly.

leptun replied:
Flash increase... That sounds wrong. What if you use a printf_P (also for pinda status, ternary evaluation to 1:0)?

leptun replied:
`printf_P(PSTR("PINDA cal status: %u\n"), calibration_status_pinda() ? 1 : 0);` sounds correct to me. I hope it's optimal. In any case, I'll always trade flash for ram. Issues caused by ram can be very deceiving, while a lack of flash means the build will fail (obviously trivial to debug).

leptun replied:
Please don't commit to this. I have an awesome improvement. I'll squash this commit into the original Pinda status commit

leptun replied:
See https://github.com/prusa3d/Prusa-Firmware/pull/3378/commits/aa7aefd6f844c470f1616d9a88632dd0827417ef

leptun committed:
Optimise PINDA cal status Serial print

"PINDA cal status:" is always printed no matter what cal_status is so we can pull that out of the conditional statment.

cal_status is also a boolean, lets just print it directly. Its a simpler code.

Saves 32 bytes of flash and 22 bytes of SRAM

Change serial messages to PGM

leptun committed:
Make the "ln" functions no-inline.

Save 348B of flash

leptun commented:
In case of emergency, we could try to use `-fmerge-all-constants`:
> -fmerge-all-constants
Attempt to merge identical constants and identical variables.
This option implies -fmerge-constants. In addition to -fmerge-constants this considers e.g. even constant initialized arrays or initialized constant variables with integral or floating point types. Languages like C or C ++ require each variable, including multiple instances of the same variable in recursive calls, to have distinct locations, so using this option will result in non-conforming behavior.

Doing just a quick compile test resulted in 800B of savings. But as I said, only in case of emergency since it can have nasty side effects.

I wonder if this option merges duplicate PROGMEM strings. If it does, it would be amazing.

gudnimg committed:
Change SERIAL_PROTOCOL to SERIAL_PROTOCOLPGM in a few lines

Saves 12 bytes of SRAM but increases flash use by 12 bytes.

leptun commented:
@gudnimg Rebased onto latest MK3 and removed commits already merged

gudnimg committed:
Optimise code with memcpy in M600 and G0/G1

Saves 64 bytes of flash memory and 9 lines of code :)

gudnimg committed:
Merge branch 'MK3_3.12_Optimizations_extra' of https://github.com/leptun/Prusa-Firmware into MK3_3.12_Optimizations_extra

leptun commented:
Todo: Optimize `M115 U` and `M862.4`. There is a lot of duplicate code with these two

gudnimg commented:
Closing in favor of https://github.com/prusa3d/Prusa-Firmware/pull/4281
